compiler/rustc_codegen_llvm/src/va_arg.rs

```diff
@@ -271,6 +279,87 @@ fn emit_s390x_va_arg<'ll, 'tcx>(
     bx.load(val_type, val_addr, layout.align.abi)
 }
 
+fn emit_xtensa_va_arg<'ll, 'tcx>(
+    bx: &mut Builder<'_, 'll, 'tcx>,
+    list: OperandRef<'tcx, &'ll Value>,
+    target_ty: Ty<'tcx>,
+) -> &'ll Value {
+    // Implementation of va_arg for Xtensa. There doesn't seem to be an authoritative source for
+    // this, other than "what GCC does".
+    //
+    // The va_list type has three fields:
+    // struct __va_list_tag {
+    //   int32_t *va_stk; // Arguments passed on the stack
+    //   int32_t *va_reg; // Arguments passed in registers, saved to memory by the prologue.
```

Comment:
Yes, structs can be passed via registers, structs can even be split across registers and stack space, _but_ primitive types will never be split. For instance, a `u128` either has to be all in registers, or all in stack never split across both.

> The reason I'm asking is because I want to know if this is one of those varargs ABIs where extern "C" fn (c_int, ...) where only 1 arg, a c_int, is passed, is ABI-equivalent to extern "C" fn (c_int, c_int) on entry into the actual function. Some ABIs deliberately design their varargs implementation to work like that, allowing you to basically deal with those two function pointers interchangeably. Others do not allow such because all varargs are handled in a distinct way so you only would match ABI details between varargs and other functions with only fixed args by accident.

From the Xtensa ISA:

![image](https://github.com/user-attachments/assets/f87fd0ec-0156-4e20-bd78-a97055c2e5e7)

So I believe in this case calling a normal function or vararg function is the same, but like you mentioned the function then has to unpack it's arguments in the vararg case.